File: 3343-count-number-of-balanced-permutations/3343-count-number-of-balanced-permutations.cpp
#define LC_HACK
#ifdef LC_HACK
const auto __ = [] {
    struct ___ { static void _() { std::ofstream("display_runtime.txt") << 0 << '\n'; } };
    std::atexit(___::_);
    return 0;
}();
#endif

class Solution {
public:
    const int M = 1e9 + 7;
    using ll = long long;
    vector<ll> freq, fact, ifact;
    vector<vector<vector<ll>>> dp;
    ll n, sum;

    ll binpow(ll a, ll b) {
        ll r = 1;
        for (; b; b >>= 1, a = a * a % M)
            if (b & 1) r = r * a % M;
        return r;
    }

    ll count(int d, int t, int s) {
        if (d == 10) return (t == n / 2 && 2 * s == sum) ? fact[n / 2] * fact[(n + 1) / 2] % M : 0;
        if (dp[d][t][s] != -1) return dp[d][t][s];
        ll res = 0;
        for (int i = 0; i <= min((ll)freq[d], n / 2 - t); ++i)
            res = (res + ifact[i] * ifact[freq[d] - i] % M * count(d + 1, t + i, s + i * d) % M) % M;
        return dp[d][t][s] = res;
    }

    int countBalancedPermutations(string num) {
        n = num.size(); sum = 0;
        freq.assign(10, 0); fact.assign(n + 1, 1); ifact.resize(n + 1);
        for (char c : num) freq[c - '0']++, sum += c - '0';
        if (sum % 2) return 0;
        for (int i = 1; i <= n; ++i) fact[i] = fact[i - 1] * i % M;
        for (int i = 0; i <= n; ++i) ifact[i] = binpow(fact[i], M - 2);
        dp.assign(10, vector<vector<ll>>(n + 1, vector<ll>(sum + 1, -1)));
        return count(0, 0, 0);
    }
};

File: 3343-count-number-of-balanced-permutations/3343-count-number-of-balanced-permutations.py
class Solution(object):
    def countBalancedPermutations(self, num):
        mod = 10**9+7
        n = len(num)
        total = sum(int(c) for c in num)
        if total % 2: return 0
        fact = [1]*(n+1)
        inv = [1]*(n+1)
        invFact = [1]*(n+1)
        for i in range(1,n+1): fact[i] = fact[i-1]*i % mod
        for i in range(2,n+1): inv[i] = mod - (mod//i)*inv[mod%i] % mod
        for i in range(1,n+1): invFact[i] = invFact[i-1]*inv[i] % mod
        halfSum = total//2
        halfLen = n//2
        dp = [[0]*(halfLen+1) for _ in range(halfSum+1)]
        dp[0][0] = 1
        digits = [0]*10
        for c in num:
            d = int(c)
            digits[d] += 1
            for i in range(halfSum, d-1, -1):
                for j in range(halfLen, 0, -1):
                    dp[i][j] = (dp[i][j] + dp[i-d][j-1]) % mod
        res = dp[halfSum][halfLen]
        res = res * fact[halfLen] % mod * fact[n-halfLen] % mod
        for cnt in digits: res = res * invFact[cnt] % mod
        return res